rts/at/annotation/README.md

```diff
@@ -0,0 +1,59 @@
+OVERVIEW
+--------
+A example that demonstrates how to use [REST-AT annotations](https://github.com/jbosstm/narayana/tree/main/rts/at/tx/src/main/java/org/jboss/jbossts/star/annotation) to define transactional microservices.
+
+USAGE
+--------------------
+
+Prior to running the example make sure that the [REST-AT coordinator is deployed](../README.md#usage).
+
+Once REST-AT coordinator is deployed build and run the annotation example application to run on port 8082, 8083 and 8084 as it in configured in application.properties
+    
+    cd <narayana-repo>/rts/at/annotation/
+    mvn clean package -DskipTests
+    java -jar flight-service/target/quarkus-app/quarkus-run.jar &
+    java -jar hotel-service/target/quarkus-app/quarkus-run.jar &
+    java -jar trip-service/target/quarkus-app/quarkus-run.jar &
+
+Repeat the same step for hotel-service and trip-service, this will start three micro services on three different ports.
+
+Book a trip within a transaction, refer to the [`bookTrip` method in the trip service](./trip-service/src/main/java/io/narayana/sra/demo/api/TripController.java) to see how it is annotated. The following curl command will invoke the resource:this starts a transaction and you can verify it is running using
+    
+    curl -XPOST http://localhost:8082/trip/book?hotelName=Rex
+
+Since the `bookTrip` resource is annotated with `@SRA(value = SRA.Type.REQUIRED, end = false)`, the curl request will start a new REST-AT transaction and you can verify that it is running by querying the coordinator:  `curl http://localhost:8080/rest-at-coordinator/tx/transaction-manager`.
+
+EXPECTED OUTPUT
+---------------
+    SRA: 0_ffff0a4cf0b6_-568c278f_6458bef8_7fb: Updating hotel participant state to: TransactionPrepared
+    SRA: 0_ffff0a4cf0b6_-568c278f_6458bef8_7fb: Updating trip participant state to: TransactionPrepared
+    SRA: 0_ffff0a4cf0b6_-568c278f_6458bef8_7fb: Updating trip participant state to: TransactionCommitted
+    SRA: 0_ffff0a4cf0b6_-568c278f_6458bef8_7fb: Updating hotel participant state to: TransactionCommitted
```

Comment:
I am working on flight participants, as it is giving error 500, I am wondering why the other APIs are not rolling back.

Reply:
@mmusgrov It seems that the problem is from [sra](https://github.com/jbosstm/narayana/tree/main/rts/sra) it self.
I just tested it and found the same behaviour, where after first call inside a trip to hotel response is success and the next call i.e flight service the response is failed with RESTEASY003765: Response is closed error. I tried to swap the service call first flight success and the next hotel service failed with the same exception. Its seems that the second service call is failing and also the status is not rolling back. So this means the SRA is not working perfectly.

Could you please check and let me know ? I really need your help on this.
Thanks.

Reply:
The flight isn't getting booked because the README does not specify a flight to be included in the booking. You need to tell the user to type:

```
curl -XPOST http://localhost:8082/trip/book?hotelName=Rex\&flightNumber=123
```

The other issue is that the call from the trip service to the hotel service to book the room is propagating the transaction context to a method annotated with `@SRA(SRA.Type.REQUIRED)` but the default value of the end attribute is `true` so the transaction is committed after booking the room. Therefore when you try to book the flight no transaction is active so it doesn't get included in the trip booking. You need call methods annotated with `@SRA(value = SRA.Type.REQUIRED, end = false)` until you have finished booking each component of the trip and only then end the transaction. My advice is that you add a method to the trip controller which ends the transaction and add a step to the README telling the reader to invoke that endpoint to finalise the booking.

The `RESTEASY003765` error is of more concern, but if you can address my previous two points first then we can look to see if you are still getting this error (I'm just guessing but maybe for certain JAX-RS response codes we aren't able to read the entity body but let's fix the quickstart first).

Reply:
Another option to fix the flight booking issue is to just remove the `end=false` attribute from the `bookTrip` method. But if you do do that remember to update the comment before the method declaration.

Reply:
I fixed and merged the RESTEASY003765 problem (https://github.com/jbosstm/narayana/pull/2153). But if you sort out the other two problems then it shouldn't go down that route.

Reply:
> The flight isn't getting booked because the README does not specify a flight to be included in the booking. You need to tell the user to type:
> 
> ```
> curl -XPOST http://localhost:8082/trip/book?hotelName=Rex\&flightNumber=123
> ```
> 
> The other issue is that the call from the trip service to the hotel service to book the room is propagating the transaction context to a method annotated with `@SRA(SRA.Type.REQUIRED)` but the default value of the end attribute is `true` so the transaction is committed after booking the room. Therefore when you try to book the flight no transaction is active so it doesn't get included in the trip booking. You need call methods annotated with `@SRA(value = SRA.Type.REQUIRED, end = false)` until you have finished booking each component of the trip and only then end the transaction. My advice is that you add a method to the trip controller which ends the transaction and add a step to the README telling the reader to invoke that endpoint to finalise the booking.
> 
> The `RESTEASY003765` error is because we are [reading the entity twice](https://github.com/jbosstm/narayana/blob/main/rts/at/tx/src/main/java/org/jboss/jbossts/star/client/SRAClient.java#L551C1-L554) but this will not manifest if you fix the other errors.

Thanks for the suggestion @mmusgrov ,  I tried  adding `@SRA(value = SRA.Type.REQUIRED, end = false)`  in hotel service and tested, working perfectly all 3 service for success scenario. I tried testing failure scenario in hotel service by sending the 500 status the transaction is rolledback. Seems working perfectly. I will update the PR soon.

Thanks.

Reply:
> Another option to fix the flight booking issue is to just remove the `end=false` attribute from the `bookTrip` method. But if you do do that remember to update the comment before the method declaration.

Yes instead of adding method which ends the transaction this is best option I tried and working.

Reply:
> I fixed and merged the RESTEASY003765 problem ([jbosstm/narayana#2153](https://github.com/jbosstm/narayana/pull/2153)). But if you sort out the other two problems then it shouldn't go down that route.

Thanks for the update

Reply:
Hi @mmusgrov I've updated the changes, could you please check and resolve the conversation?
Thanks.

Reply:
> > The flight isn't getting booked because the README does not specify a flight to be included in the booking. You need to tell the user to type:
> > ```
> > curl -XPOST http://localhost:8082/trip/book?hotelName=Rex\&flightNumber=123
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The other issue is that the call from the trip service to the hotel service to book the room is propagating the transaction context to a method annotated with `@SRA(SRA.Type.REQUIRED)` but the default value of the end attribute is `true` so the transaction is committed after booking the room. Therefore when you try to book the flight no transaction is active so it doesn't get included in the trip booking. You need call methods annotated with `@SRA(value = SRA.Type.REQUIRED, end = false)` until you have finished booking each component of the trip and only then end the transaction. My advice is that you add a method to the trip controller which ends the transaction and add a step to the README telling the reader to invoke that endpoint to finalise the booking.
> > The `RESTEASY003765` error is because we are [reading the entity twice](https://github.com/jbosstm/narayana/blob/main/rts/at/tx/src/main/java/org/jboss/jbossts/star/client/SRAClient.java#L551C1-L554) but this will not manifest if you fix the other errors.
> 
> Thanks for the suggestion @mmusgrov , I tried adding `@SRA(value = SRA.Type.REQUIRED, end = false)` in hotel service and tested, working perfectly all 3 service for success scenario. I tried testing failure scenario in hotel service by sending the 500 status the transaction is rolledback. Seems working perfectly. I will update the PR soon.
> 
> Thanks.

`@SRA(value = SRA.Type.REQUIRED, end = false)`  Is needed to add to flight service as well because if flight service fails it has to roll back, otherwise it is committing the transaction.

Reply:
Hi @mmusgrov I've updated the changes, could you please check and resolve the conversation?
Thanks.

Reply:
> > Another option to fix the flight booking issue is to just remove the `end=false` attribute from the `bookTrip` method. But if you do do that remember to update the comment before the method declaration.
> 
> Yes instead of adding method which ends the transaction this is best option I tried and working.

Yes but now since the transaction is still active when the `bookTrip` method returns the booking status is reported as `PROVISIONAL`, recall that the meaning of `... end=true` is that the transaction commits after the business method finishes:

```
{
  "cancelPending": false,
  "details": [
    {
      "cancelPending": false,
      "details": [],
      "encodedId": "",
      "id": "",
      "name": "",
      "quantity": 0,
      "status": "PROVISIONAL",
      "type": ""
    },
    {
      "cancelPending": false,
      "details": [],
      "encodedId": "",
      "id": "",
      "name": "",
      "quantity": 0,
      "status": "PROVISIONAL",
      "type": ""
    }
  ],
  "encodedId": "0_ffffac1182ac_-46776e66_64affd81_d8",
  "id": "0_ffffac1182ac_-46776e66_64affd81_d8",
  "name": "Aggregate Booking",
  "quantity": 1,
  "status": "PROVISIONAL",
  "type": "Trip"
}
```

When the bookTrip method completes the JAX-RS `ContainerResponseFilter` runs (take another look at ServerSRAFilter.java) and that sends a request to the coordinator asking it to commit the transaction. The coordinator will then call back into each service that participated in the transaction to tell them to first `prepare` and to then `commit`.

So reporting the booking status before this 2PC protocol runs is going to confuse the reader and put him off from using REST-AT.

So you need to:

1. Store the booking in the BookingStore, each service extends BookingStore so you just need to call `add()` from the bookTrip business method, take a look at how the flight and hotel services do it.
2. Handle the prepare and commit callbacks and update the booking status, see the callback handler `protected SRAStatus updateParticipantState(SRAStatus status, String activityId)` in the TripController.
3. Add an endpoint in the `TripController` to GET the booking and that should report the aggregate status as `CONFIRMED`.

I need to prepare for a release of Narayana 7.0.0.Final tomorrow so I can't fully code it up but if you have any questions then ask me. But basically you need do the following:

```
diff --git a/rts/at/annotation/trip-service/src/main/java/io/narayana/sra/demo/service/TripService.java b/rts/at/annotation/trip-service/src/main/java/io/narayana/sra/demo/service/TripService.java
index 42032773c..70a95c095 100644
--- a/rts/at/annotation/trip-service/src/main/java/io/narayana/sra/demo/service/TripService.java
+++ b/rts/at/annotation/trip-service/src/main/java/io/narayana/sra/demo/service/TripService.java
@@ -23,6 +23,14 @@ public class TripService extends BookingStore{
         this.lraClient = new SRAClient();
     }
 
+    public Booking book(String bid, String hotel, Integer beds) {
+        Booking booking = new Booking(bid, hotel, beds, "Hotel");
+
+        add(booking);
+
+        return booking;
+    }
```

and

```
+++ b/rts/at/annotation/trip-service/src/main/java/io/narayana/sra/demo/api/TripController.java                                                                                                                                               
@@ -98,9 +98,21 @@ public class TripController extends SRAParticipant {                                                                                                                                                                       
                                                                                                                                                                                                                                              
         Booking tripBooking = new Booking(sraId, "Trip", hotelBooking, flightBooking);                                                                                                                                                       
                                                                                                                                                                                                                                              
+        tripService.add(tripBooking);                                                                                                                                                                                                        
+                                                                                                                                                                                                                                             
         return Response.status(Response.Status.CREATED).entity(tripBooking).build();                                                                                                                                                         
     }                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                              
+    @GET                                                                                                                                                                                                                                     
+    @Path("")                                                                                                                                                                                                                                
+    @Produces(MediaType.APPLICATION_JSON)                                                                                                                                                                                                    
+    @SRA(SRA.Type.NOT_SUPPORTED)                                                                                                                                                                                                             
+    public Response getBooking(@QueryParam("bid") String bookingId) throws NotFoundException {                                                                                                                                                   
+        Booking tripBooking = tripService.get(bookingId);                                                                                                                                                                                        
+                                                                                                                                                                                                                                             
+        return Response.status(Response.Status.OK).entity(tripBooking).build();                                                                                                                                                              
+    }
+
     @GET
     @Path("/status")
     @Produces(MediaType.APPLICATION_JSON)
@@ -150,6 +162,22 @@ public class TripController extends SRAParticipant {
     @Override
     protected SRAStatus updateParticipantState(SRAStatus status, String activityId) {
         System.out.printf("SRA: %s: Updating trip participant state to: %s%n", activityId, status);
+        Booking tripBooking = tripService.get(activityId);
+
+        switch (status) {
+            case TransactionPrepared:
+                tripBooking.setConfirming();
+                break;
+            case TransactionCommitted:
+                tripBooking.setConfirmed();
+                break;
+            case TransactionRolledBack:
+                tripBooking.setCanceled();
+                break;
+            default:
+                break;
+        }
+
         return status;
     }
```

You'll also need to merge in the booking status of the booking line items (flight and hotel) by calling the `Booking.merge(Booking booking)` method, you can get the flight and hotel booking entities by calling the `/info/{bookingId}` endpoints of the flight and booking controllers.

Reply:
And for bonus points see if you figure out why the `TripController.bookHotel() `method isn't decoding the booking response correctly (`response.readEntity(Booking.class);`). I checked in a debugger and the HotelController is sending back a fully populated booking object so the problem will be to do with how we've configured the JAX-RS message body readers and writers.